nupic/research/frameworks/vernon/mixins/centroid_context.py

```diff
@@ -109,16 +172,174 @@ def _train_centroid(data):
     return _train_centroid
 
 
-def infer_centroid(contexts):
+def cluster_centroid_while_training(clusters, contexts, n_samples_per_context,
+                                    subindices=None):
+    """
+    Returns a function that takes a batch of training examples and returns a function
+    that performs a clustering procedure to determine the appropriate context vector.
+    The resulting context vector returned by the function is either a) an existing
+    context vector in `contexts` or b) simply the centroid of the batch.
+
+    :param clusters: List of Torch Tensors where the item at position i gives the
+                     exemplars representing cluster i
+    :param contexts: List containing a single Torch Tensor in which row i gives the ith
+                     context vector
+    :param n_samples_per_context: List of ints where entry i gives the number of
+                                  samples used to compute the ith context (i.e.,
+                                  `contexts[0][i]`)
+    :param subindices: List/Tensor/Array that can index contexts to select subindices;
+                       optional
+    """
+
+    def _cluster(data):
+        nonlocal contexts
```

Comment:
Do you need this? I'm not familiar with `nonlocal`. Can you just do a closure instead?